lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,653 @@
+  for (auto it : tappedArgs)
+    LLVM_DEBUG(llvm::dbgs() << "- " << it.first << ": " << it.second << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "Tapped ops:\n");
+  for (auto it : tappedOps)
+    LLVM_DEBUG(llvm::dbgs() << "- " << it.first << ": " << *it.second << "\n");
+#endif
+
+  // Process each black box independently.
+  for (auto blackBox : modules) {
+    LLVM_DEBUG(llvm::dbgs() << "Generating impls for "
+                            << blackBox.extModule.getName() << "\n");
+
+    // As a first step, gather a list of all absolute paths to instances of this
+    // black box.
+    auto paths = instanceGraph.getAbsolutePaths(blackBox.extModule);
+#ifndef NDEBUG
+    for (auto path : paths)
+      LLVM_DEBUG(llvm::dbgs() << "- " << path << "\n");
+#endif
+
+    // Go through the port annotations of the tap module and generate a
+    // hierarchical path for each.
+    portWiring.clear();
+    portWiring.reserve(blackBox.portAnnos.size());
+    for (auto portAnno : blackBox.portAnnos) {
+      processAnnotation(portAnno, blackBox, strings, instanceGraph);
+    }
+
+#ifndef NDEBUG
+    LLVM_DEBUG(llvm::dbgs() << "- Wire up as follows:\n");
+    for (auto wiring : portWiring) {
+      LLVM_DEBUG(llvm::dbgs() << "- Port " << wiring.portNum << ":\n");
+      for (auto path : wiring.prefices) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "  - " << path << "." << wiring.suffix << "\n");
+      }
+    }
+#endif
+
+    // Now we have an awkward mapping problem. We have multiple data tap module
+    // instances, which reference things in modules that in turn have multiple
+    // instances. This is a side-effect of how Grand Central annotates things on
+    // modules rather than instances. (However in practice these will have a
+    // one-to-one correspondence due to CHIRRTL having fully uniquified
+    // instances.) To solve this issue, create a dedicated implementation for
+    // every data tap instance, and among the possible targets for the data taps
+    // choose the one with the shortest relative path to the data tap instance.
+    ImplicitLocOpBuilder builder(blackBox.extModule->getLoc(),
+                                 blackBox.extModule);
+    unsigned implIdx = 0;
+    for (auto path : paths) {
+      builder.setInsertionPointAfter(blackBox.extModule);
+
+      // Get the list of ports from the original extmodule, and update the
+      // annotations such that they no longer contain any data/mem taps.
+      auto ports = blackBox.extModule.getPorts();
+      for (auto port : llvm::zip(ports, blackBox.filteredPortAnnos)) {
+        std::get<0>(port).annotations = std::get<1>(port);
+      }
+
+      // Create a new firrtl.module that implements the data tap.
+      auto name =
+          StringAttr::get(&getContext(), (Twine(blackBox.extModule.getName()) +
+                                          "_impl_" + llvm::utostr(implIdx++))
+                                             .str());
```

Comment:
You can drop the .str() now, and use `Twine(implIdx++)` which is more efficient.

Reply:
Done. That change on `StringAttr` is pretty handy.